hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/impl/SqlClientConnection.java

```diff
@@ -362,52 +367,86 @@ private SqlConnection client() {
-				.onSuccess( v -> LOG.tracef( "Transaction committed: %s", transaction ) )
-				.onFailure( v -> LOG.errorf( "Failed to commit transaction: %s", transaction ) )
-				.toCompletionStage()
-				.whenComplete( this::clearTransaction );
+		return connection.transaction()
+				.commit()
+				.onSuccess( v -> LOG.tracef( "Transaction committed: %s", connection.transaction() ) )
+				.onFailure( throwable -> LOG.errorf( "Failed to commit transaction: %s", throwable.getMessage() ) )
+				.toCompletionStage();
 	}
 
 	@Override
 	public CompletionStage<Void> rollbackTransaction() {
-		return transaction.rollback()
-				.onFailure( v -> LOG.errorf( "Failed to rollback transaction: %s", transaction ) )
-				.onSuccess( v -> LOG.tracef( "Transaction rolled back: %s", transaction ) )
-				.toCompletionStage()
-				.whenComplete( this::clearTransaction );
+		if ( isTransactionInProgress() ) {
+			return connection.transaction()
+					.rollback()
+					.onFailure( throwable -> LOG.errorf( "Failed to rollback transaction: %s", throwable.getMessage() ) )
+					.onSuccess( v -> LOG.tracef( "Transaction rolled back: %s", connection.transaction() ) )
+					.toCompletionStage();
+		}
+		LOG.trace( "No transaction found to roll back" );
+		return voidFuture();
 	}
 
 	@Override
 	public CompletionStage<Void> close() {
-		if ( transaction != null ) {
-			throw new IllegalStateException( "Connection being closed with a live transaction associated to it" );
-		}
-		return connection.close()
-				.onSuccess( event -> LOG.tracef( "Connection closed: %s", connection ) )
-				.onFailure( v -> LOG.errorf( "Failed to close a connection: %s", connection ) )
-				.toCompletionStage();
-	}
-
-	private void setTransaction(Transaction tx) {
-		transaction = tx;
+		// We can probably skip the validation if the connection is already closed...but, you never know
+		return validateLiveTransaction( "closing the connection" )
+				.handle( CompletionStages::handle )
+				.thenCompose( validationHandler -> supplyStage( () -> closed
+						? voidFuture().thenAccept( v -> LOG.trace( "Connection already closed" ) )
+						: connection.close().toCompletionStage() )
+						.handle( CompletionStages::handle )
+						.thenCompose( closeConnectionHandler -> {
+							if ( closeConnectionHandler.hasFailed() ) {
+								if ( validationHandler.hasFailed() ) {
+									// Error closing the connection, include the validation error
+									closeConnectionHandler.getThrowable()
+											.addSuppressed( validationHandler.getThrowable() );
+								}
+								return closeConnectionHandler.getResultAsCompletionStage();
+							}
+							if ( !closed ) {
+								closed = true;
+								LOG.tracef( "Connection closed: %s", connection );
+							}
+							else {
+								LOG.tracef( "Connection was already closed: %s", connection );
+							}
+							// Connection closed, return the result of the validation
+							return validationHandler.getResultAsCompletionStage();
+						} )
+				);
```

Comment:
There's... a lot to unpack here, and I wish we had an easier way to express all this. IIUC this is mostly to suppress exceptions?

Anyway, as I can see, this does what we want.

Reply:
Yes, it's basically a try-catch-finally block. We need to make sure that we are still closing the connection when there's a validation error.  If closing the connection fails, I add the validation error as a suppressed exception. This way we don't hide errors. 

Maybe if there's a validation error, but I can still close the connection, I should mention in the error message that the connection was closed